client/login/magic-login/handle-emailed-link-form.jsx

```diff
@@ -40,42 +76,75 @@ class HandleEmailedLinkForm extends React.Component {
 		this.props.fetchMagicLoginAuthenticate( this.props.emailAddress, this.props.token, this.props.tokenTime );
 	};
 
-	componentWillMount() {
-		const { emailAddress, token, tokenTime } = this.props;
-
-		if ( emailAddress && emailValidator.validate( emailAddress ) && token && tokenTime ) {
-			return;
+	// Lifted from `blocks/login`
+	// @TODO unify
+	handleValidToken = () => {
+		if ( ! this.props.twoFactorEnabled ) {
+			this.rebootAfterLogin();
+		} else {
+			page( login( {
+				isNative: true,
+				// If no notification is sent, the user is using the authenticator for 2FA by default
+				twoFactorAuthType: this.props.twoFactorNotificationSent.replace( 'none', 'authenticator' )
+			} ) );
 		}
+	};
```

Comment:
Handlers like this one only depend on Redux-connected props — not own props — for their logic. To me, that further gives us reason to treat this as an action and move it out of the component. Same goes for `rebootAfterLogin` — actually, they're intertwined.